gluonnlp/model/highway.py

```diff
@@ -0,0 +1,108 @@
+# coding: utf-8
+
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Highway layer."""
+
+from __future__ import absolute_import
+from __future__ import print_function
+
+__all__ = ['Highway']
+
+from mxnet import gluon
+from mxnet.gluon import nn
+
+
+class Highway(gluon.Block):
+    r"""Highway network.
+
+    We implemented the highway network proposed in the following work::
+
+        @article{srivastava2015highway,
+          title={Highway networks},
+          author={Srivastava, Rupesh Kumar and Greff, Klaus and Schmidhuber, J{\"u}rgen},
+          journal={arXiv preprint arXiv:1505.00387},
+          year={2015}
+        }
+
+    A Highway layer is defined as below:
+
+    .. math::
+        y = (1 - t) * x + t * f(A(x))
+
+    which is a gated combination of a linear transform and a non-linear transform of its input,
+    where :math:`x` is the input tensor, :math:`A` is a linear transformer,
+    :math:`f` is an element-wise non-linear transformer,
+    and :math:`t` is an element-wise transform gate, and :math:`t` refers to carry gate.
+
+    Parameters
+    ----------
+    input_size : int
+        The dimension of the input tensor.  We assume the input has shape ``(batch_size,
+        input_size)``.
+    num_layers : int
+        The number of highway layers to apply to the input.
+    activation : nn.Activation
+        The non-linear activation function.
+    """
+
+    def __init__(self,
+                 input_size,
+                 num_layers,
+                 activation=nn.Activation('relu'),
+                 **kwargs):
+        super(Highway, self).__init__(**kwargs)
+        self._input_size = input_size
+        self._num_layers = num_layers
+
+        with self.name_scope():
+            self.hnet = nn.HybridSequential()
+            with self.hnet.name_scope():
+                for _ in range(self._num_layers):
+                    self.hnet.add(nn.Dense(units=self._input_size * 2, in_units=self._input_size,
```

Comment:
Would it make the code easier to read/maintain when using two separate Dense layers instead of a single one with twice the number of units and then splitting the units in half during forward? Then you can consider to use the `bias_initializer` parameter for the Dense layer that is used as Transform gate. It could be set by default to `mx.initializer.Constant(-1)` and users could overwrite it by passing an option like `transform_gate_bias_initializer` to `Highway`.

Reply:
The `g` would be a non-linearity, the input projected and combined with the non-transform follows the definition of `g`. We are now using a highway bias to satisfy the requirement of highway initialization.